video/out/vo_dmabuf_wayland.c

```diff
@@ -0,0 +1,359 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <unistd.h>
+
+#include "config.h"
+
+#ifdef HAVE_VAAPI
+#include <va/va_drmcommon.h>
+#include <libavutil/hwcontext_drm.h>
+#endif
+
+#include "mpv_talloc.h"
+#include "common/global.h"
+#include "vo.h"
+#include "video/mp_image.h"
+
+#include "gpu/hwdec.h"
+#include "gpu/video.h"
+
+#include "video/vaapi.h"
+#include "present_sync.h"
+#include "wayland_common.h"
+#include "generated/wayland/linux-dmabuf-unstable-v1.h"
+#include "generated/wayland/viewporter.h"
+#include "wlbuf_pool.h"
+
+struct priv {
+    struct mp_log *log;
+    struct ra_ctx *ctx;
+    struct mpv_global *global;
+    struct ra_hwdec_ctx hwdec_ctx;
+    int events;
+
+    struct wl_shm_pool *solid_buffer_pool;
+    struct wl_buffer *solid_buffer;
+    struct wlbuf_pool *wlbuf_pool;
+    bool want_reset;
+    uint64_t reset_count;
+
+#ifdef HAVE_VAAPI
+    VADisplay display;
+#endif
+};
+
+#ifdef HAVE_VAAPI
+static uintptr_t vaapi_key_provider(struct mp_image *src)
+{
+    return va_surface_id(src);
+}
+
+/* va-api dmabuf importer */
+static bool vaapi_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_entry* entry,
+                                  struct zwp_linux_buffer_params_v1 *params)
+{
+    struct priv *p = entry->vo->priv;
+    VADRMPRIMESurfaceDescriptor desc;
+    bool dmabuf_imported = false;
+    /* composed has single layer */
+    int layer_no = 0;
+    VAStatus status = vaExportSurfaceHandle(p->display, entry->key, VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2,
+                                            VA_EXPORT_SURFACE_COMPOSED_LAYERS | VA_EXPORT_SURFACE_READ_ONLY, &desc);
+    if (status == VA_STATUS_ERROR_INVALID_SURFACE) {
+        MP_VERBOSE(entry->vo, "VA export to composed layers not supported.\n");
+    } else if (!vo_wayland_supported_format(entry->vo, desc.layers[0].drm_format, desc.objects[0].drm_format_modifier)) {
+        MP_VERBOSE(entry->vo, "%s(%016lx) is not supported.\n",
+                   mp_tag_str(desc.layers[0].drm_format), desc.objects[0].drm_format_modifier);
+    } else if (CHECK_VA_STATUS(entry->vo, "vaExportSurfaceHandle()")) {
+        entry->drm_format = desc.layers[layer_no].drm_format;
+        for (int plane_no = 0; plane_no < desc.layers[layer_no].num_planes; ++plane_no) {
+            int object = desc.layers[layer_no].object_index[plane_no];
+            uint64_t modifier = desc.objects[object].drm_format_modifier;
+            zwp_linux_buffer_params_v1_add(params, desc.objects[object].fd, plane_no, desc.layers[layer_no].offset[plane_no],
+                                           desc.layers[layer_no].pitch[plane_no], modifier >> 32, modifier & 0xffffffff);
+        }
+        dmabuf_imported = true;
+    }
+
+    /* clean up descriptor */
+    for (int i = 0; i < desc.num_objects; i++) {
+        close(desc.objects[i].fd);
+        desc.objects[i].fd = 0;
```

Comment:
This segfaults if `VA_STATUS_ERROR_INVALID_SURFACE` is hit above. Could just return false on that line or add a goto to skip this?